### PR TITLE
JS API: Fixes & maintenance: commands new bases (ED-2084)

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: assets/dev/js/editor/base/command-editor-internal.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import CommandEditor from './command-editor';
```
Comment (iNewLegend) [inline assets/dev/js/editor/base/command-editor-internal.js:1]: test

File: assets/dev/js/editor/base/command-editor.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import Command from 'elementor-api/modules/command';
```
Comment (iNewLegend) [inline assets/dev/js/editor/base/command-editor.js:1]: test

File: core/common/assets/js/api/core/commands.js [hunk 3]
```diff
@@ -325,17 +399,44 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+	validateInstance( instance, command ) {
+		if ( ! ( instance instanceof Command ) ) {
+			this.error( `invalid instance, command: '${ command }' ` );
+		}
+
+		const component = this.getComponent( command );
+
+		if ( component !== instance.component ) {
+			if ( $e.devTools ) {
+				$e.devTools.log.error( `Command: '${ command }' registerArgs.component: '${ instance.component.getNamespace() }' while current: '${ component.getNamespace() }'` );
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:431]: After time, the error not understand able.

File: core/common/assets/js/api/modules/command-callback.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import Command from './command';
```
Comment (iNewLegend) [inline core/common/assets/js/api/modules/command-callback.js:1]: test

File: core/common/assets/js/api/modules/command-internal.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import Command from './command';
```
Comment (iNewLegend) [inline core/common/assets/js/api/modules/command-internal.js:1]: test

File: core/common/assets/js/api/modules/command.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import CommandBase from './command-base';
```
Comment (iNewLegend) [inline core/common/assets/js/api/modules/command.js:1]: test

File: assets/dev/js/editor/base/command-editor-internal.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+export default class CommandInternal extends CommandEditor {
```
Comment (iNewLegend) [inline assets/dev/js/editor/base/command-editor-internal.js:3]:
```suggestion
export default class CommandEditorInternal extends CommandEditor {
```

File: core/common/assets/js/api/core/commands.js [hunk 8]
```diff
@@ -325,17 +402,44 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
 		delete this.current[ container ];
 		delete this.currentArgs[ container ];
 	}
 
+	validateInstance( instance, command ) {
+		if ( ! ( instance instanceof Command ) ) {
+			this.error( `invalid instance, command: '${ command }' ` );
+		}
+
+		const component = this.getComponent( command );
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:430]:
```suggestion
		const currentComponent = this.getComponent( command );
```

File: core/common/assets/js/api/index.js [hunk 9]
```diff
@@ -17,6 +17,12 @@ import Shortcuts from './core/shortcuts';
 import * as hookData from './modules/hooks/data/';
 import * as hookUI from './modules/hooks/ui';
 
+/**
+ * @typedef {{}} Args
+ * @property {Container} container
+ * @property {Array.<Container>} containers
+ */
+
```
Comment (iNewLegend) [inline core/common/assets/js/api/index.js:25]: Revert this change api/index should not know about the existence of container(s).

File: core/common/assets/js/api/modules/command-base.js [hunk 10]
```diff
@@ -1,16 +1,29 @@
 import ArgsObject from 'elementor-assets-js/modules/imports/args-object';
 
 export default class CommandBase extends ArgsObject {
+	static registerArgs = {};
```
Comment (iNewLegend) [inline core/common/assets/js/api/modules/command-base.js:4]: JSDOC

File: core/common/assets/js/api/modules/command-base.js [hunk 11]
```diff
@@ -20,14 +33,20 @@ export default class CommandBase extends ArgsObject {
...
 		super( args );
 
+		if ( 0 === $e.commands.constructor.trace.length ) {
+			this.constructor.registerArgs = args;
+
+			return;
+		}
+
 		// Acknowledge self about which command it run.
-		this.currentCommand = commandsAPI.getCurrentLast();
```
Comment (iNewLegend) [inline core/common/assets/js/api/modules/command-base.js:27]: BC.

File: core/common/assets/js/api/core/commands.js [hunk 12]
```diff
@@ -240,22 +240,36 @@ export default class Commands extends CommandsBackwardsCompatibility {
 		return this.currentTrace[ 0 ];
 	}
 
+	validateRun( command, args = {} ) {
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:243]: JSDOC.

File: core/common/assets/js/api/modules/component-base.js [hunk 13]
```diff
@@ -15,6 +18,8 @@ export default class ComponentBase extends elementorModules.Module {
 
 		this.defaultRoute = '';
 		this.currentTab = '';
+
+		this.isRegistered = false;
```
Comment (iNewLegend) [inline core/common/assets/js/api/modules/component-base.js:22]: remove.

**Comment by matipojo** [comment]
It will be merged in the future.
